library/core/src/macros/mod.rs

```diff
@@ -1546,6 +1546,29 @@ pub(crate) mod builtin {
         /* compiler built-in */
     }
 
+    /// Unstable placeholder for type ascription.
+    #[rustc_builtin_macro]
+    #[unstable(
+        feature = "type_ascription",
+        issue = "23416",
+        reason = "placeholder syntax for type ascription"
+    )]
+    #[cfg(not(bootstrap))]
+    pub macro type_ascribe($expr:expr, $ty:ty) {
+        /* compiler built-in */
+    }
+
+    /// Unstable placeholder for type ascription.
+    #[unstable(
+        feature = "type_ascription",
+        issue = "23416",
+        reason = "placeholder syntax for type ascription"
+    )]
+    #[cfg(bootstrap)]
+    pub macro type_ascribe($expr:expr, $ty:ty) {
+        $expr: $ty
+    }
```

Comment:
Is `type_ascribe` for bootstrap necessary?

```suggestion
    /// Unstable placeholder for type ascription.
    #[rustc_builtin_macro]
    #[unstable(
        feature = "type_ascription",
        issue = "23416",
        reason = "placeholder syntax for type ascription"
    )]
    pub macro type_ascribe($expr:expr, $ty:ty) {
        /* compiler built-in */
    }
```

Reply:
I don't think it really is, I just did it because it was easy. Probably makes more sense to remove it.